src/shared/watchdog.c

```diff
@@ -17,11 +17,55 @@
 static int watchdog_fd = -1;
 static char *watchdog_device = NULL;
 static usec_t watchdog_timeout = USEC_INFINITY;
+static usec_t watchdog_pretimeout = USEC_INFINITY;
 static usec_t watchdog_last_ping = USEC_INFINITY;
 
-static unsigned int usec_to_sec(usec_t val) {
+/* Starting from kernel version 4.5, the maximum allowable watchdog timeout is
+ * UINT_MAX/1000U seconds (since internal calculations are done in milliseconds).
+ */
+#define WATCHDOG_TIMEOUT_MAX_SEC (UINT_MAX/1000U)
+
+static unsigned usec_to_sec(usec_t val) {
         usec_t t = DIV_ROUND_UP(val, USEC_PER_SEC);
-        return (unsigned int) t >= (UINT_MAX / 1000) ? (UINT_MAX / 1000) : t; /* Saturate to watchdog max */
+        return (unsigned) t >= WATCHDOG_TIMEOUT_MAX_SEC ? WATCHDOG_TIMEOUT_MAX_SEC : t; /* Saturate to watchdog max */
+}
+
+static int update_pretimeout(void) {
+        int r;
+
+        if (watchdog_fd < 0)
+                return 0;
+
+        if (watchdog_pretimeout == USEC_INFINITY || watchdog_timeout == USEC_INFINITY)
+                return 0;
+        else if (usec_to_sec(watchdog_pretimeout) >= usec_to_sec(watchdog_timeout)) {
+                unsigned t_sec = usec_to_sec(watchdog_timeout);
+                unsigned pt_sec = usec_to_sec(watchdog_pretimeout);
+                log_error("Cannot set watchdog pretimeout to %us (%s watchdog timeout of %us)",
+                          pt_sec, pt_sec == t_sec ? "same as" : "longer than", t_sec);
+                return -EINVAL;
+        } else {
+                char buf[FORMAT_TIMESPAN_MAX];
+                unsigned sec = usec_to_sec(watchdog_pretimeout);
+
+                r = ioctl(watchdog_fd, WDIOC_SETPRETIMEOUT, &sec);
+                if (r < 0) {
+                        /* EOPNOTSUPP means the watchdog does not support pretimeouts */
+                        log_full(ERRNO_IS_NOT_SUPPORTED(errno) ? LOG_INFO : LOG_WARNING,
+                                 "Failed to set pretimeout to %us: %m", sec);
+                        return (ERRNO_IS_NOT_SUPPORTED(errno) ? 0 : -errno);
+                }
+
+                /* The set ioctl does not return the actual value set, get it now */
+                if (ioctl(watchdog_fd, WDIOC_GETPRETIMEOUT, &sec))
+                        log_warning_errno(errno, "Failed to get pretimeout value: %m");
+
+                watchdog_pretimeout = (usec_t) sec * USEC_PER_SEC;
+                log_info("Set hardware watchdog pretimeout to %s.",
+                         format_timespan(buf, sizeof(buf), watchdog_pretimeout, 0));
+        }
+
+        return 0;
```

Comment:
shouldn't we try to turn off the pre-timeout if not set or set invalidly? is ther a sensible api for that?

Reply:
There's no API for disabling the pre-timeout (like setting the WDIOS_DISABLECARD flag stops the watchdog). Instead setting a pre-timeout of 0 disables the pre-timeout.

The only invalid pre-timeout is if the pre-timeout is longer than the timeout. In that case, we retain the previous pre-timeout value (by reporting an error and returning). This follows the same logic as the regular watchdog timeout, should we do something different here? If the watchdog timeout changes so that the pre-timeout value is now invalid, [the kernel will disable the pre-timeout](https://elixir.bootlin.com/linux/latest/source/drivers/watchdog/watchdog_dev.c#L390) and we will generate an error message (from `update_pretimeout `after the timeout is updated in `update_timeout`). 

For the "not set" case (when the pre-timeout is `USEC_INFINITY`), I assumed we can do nothing as we do for the regular timeout. If the pre-timeout is explicitly 0, then we will set 0 and the pre-timeout will be disabled.